### PR TITLE
add support for rsa createKey/encrypy/decrypt with sgx-ssl api

### PR DESCRIPTION
**Pull request by whh39**
it can create rsa key and length is 2048 or 3072
padding mode supports RSA_PKCS1_PADDING and RSA_PKCS1_OAEP_PADDING

add test case for rsa_2048/3072_encrypt/decrypt

test: passed the unit test

### CODE REVIEW

File: core/Enclave/enclave_hsm.cpp [hunk 1]
```diff
@@ -458,27 +496,32 @@ sgx_status_t enclave_rsa_encrypt(const uint8_t *cmk_blob,
                                  uint8_t keyspec,
                                  uint8_t padding_mode)
 {
```
Comment (yang8621) [inline core/Enclave/enclave_hsm.cpp:498]: parameters not checked

File: core/Enclave/enclave_hsm.cpp [hunk 2]
```diff
@@ -458,27 +496,32 @@ sgx_status_t enclave_rsa_encrypt(const uint8_t *cmk_blob,
                                  uint8_t keyspec,
```
Comment (yang8621) [inline core/Enclave/enclave_hsm.cpp:496]: ciphertext_len, keyspec, cmk_blob_size not used?

File: core/Enclave/enclave_hsm.cpp [hunk 3]
```diff
@@ -489,30 +532,44 @@ sgx_status_t enclave_rsa_decrypt(const uint8_t *cmk_blob, size_t cmk_blob_size,
 {
```
Comment (yang8621) [inline core/Enclave/enclave_hsm.cpp:532]: parameters not checked

File: core/Enclave/enclave_hsm.cpp [hunk 4]
```diff
@@ -489,30 +532,44 @@ sgx_status_t enclave_rsa_decrypt(const uint8_t *cmk_blob, size_t cmk_blob_size,
...
     sgx_status_t ret = SGX_ERROR_UNEXPECTED;
 
-    rsa_key_data_t *temp = (rsa_key_data_t*)cmk_blob;
-
-    BIO *bio = BIO_new_mem_buf(temp->rsa_private_key, -1);
+    // load private key
+    rsa_key_data_t *temp_pri_kry = (rsa_key_data_t*)cmk_blob;
```
Comment (yang8621) [inline core/Enclave/enclave_hsm.cpp:536]: kry?

File: core/Enclave/enclave_hsm.cpp [hunk 5]
```diff
@@ -489,30 +532,44 @@ sgx_status_t enclave_rsa_decrypt(const uint8_t *cmk_blob, size_t cmk_blob_size,
...
-    {
-        unsigned char* to;
-        RSA_private_decrypt(ciphertext_len, ciphertext, to, rsa, RSA_PKCS1_OAEP_PADDING);
-        *req_plaintext_len = strlen((char*)to);
-    }
-    else
-    {    
-        RSA_private_decrypt(ciphertext_len, ciphertext, plaintext, rsa, RSA_PKCS1_OAEP_PADDING);
+    int res;
+    if (plaintext_len == 0) {
```
Comment (yang8621) [inline core/Enclave/enclave_hsm.cpp:551]: plaintext_len refers to the length of *plaintext that is created by the normal world. If it's less than the actual size of decrypted plaintext, an error should be returned.

File: core/Enclave/enclave_hsm.cpp [hunk 6]
```diff
@@ -489,30 +532,44 @@ sgx_status_t enclave_rsa_decrypt(const uint8_t *cmk_blob, size_t cmk_blob_size,
...
+        // compute plaintext length
+        unsigned char to[RSA_size(rsa)];
```
Comment (yang8621) [inline core/Enclave/enclave_hsm.cpp:553]: either ciphertext/plaintext or from/to.

File: core/Enclave/enclave_hsm.cpp [hunk 7]
```diff
@@ -489,30 +532,44 @@ sgx_status_t enclave_rsa_decrypt(const uint8_t *cmk_blob, size_t cmk_blob_size,
...
-    }
-    else
-    {    
-        RSA_private_decrypt(ciphertext_len, ciphertext, plaintext, rsa, RSA_PKCS1_OAEP_PADDING);
+    int res;
+    if (plaintext_len == 0) {
+        // compute plaintext length
+        unsigned char to[RSA_size(rsa)]; 
+        
+        res = RSA_private_decrypt(ciphertext_len, ciphertext, to, rsa, padding_mode);
```
Comment (yang8621) [inline core/Enclave/enclave_hsm.cpp:555]: What does res mean?
RSA_private_decrypt() returns the size of the recovered plaintext.

File: core/Enclave/enclave_hsm.cpp [hunk 8]
```diff
@@ -489,30 +532,44 @@ sgx_status_t enclave_rsa_decrypt(const uint8_t *cmk_blob, size_t cmk_blob_size,
...
-    {    
-        RSA_private_decrypt(ciphertext_len, ciphertext, plaintext, rsa, RSA_PKCS1_OAEP_PADDING);
+    int res;
+    if (plaintext_len == 0) {
+        // compute plaintext length
+        unsigned char to[RSA_size(rsa)]; 
+        
+        res = RSA_private_decrypt(ciphertext_len, ciphertext, to, rsa, padding_mode);        
+        
+        *req_plaintext_len = res - 1; // delete '\0'
```
Comment (yang8621) [inline core/Enclave/enclave_hsm.cpp:557]: not understood.

File: core/App/ehsm_provider.cpp [hunk 9]
```diff
@@ -594,6 +617,39 @@ ehsm_status_t AsymmetricDecrypt(ehsm_keyblob_t *cmk,
...
+                                      cmk->keyblob,
+                                      cmk->keybloblen,
+                                      ciphertext->data,
+                                      ciphertext->datalen,
+                                      plaintext->data,
+                                      plaintext->datalen,
+                                      NULL,
+                                      cmk->metadata.keyspec,
+                                      cmk->metadata.padding_mode);
+            break;
```
Comment (syan10) [inline core/App/ehsm_provider.cpp:652]: any difference between the 2048 and 3072?

File: core/Enclave/enclave_hsm.cpp [hunk 10]
```diff
@@ -294,27 +294,59 @@ sgx_status_t enclave_create_rsa_key(uint8_t *cmk_blob,
...
             case EH_RSA_2048:
-                RSA_generate_key_ex(rsa, RSA_2048_KEY_BITS, bne, nullptr);
+                res = RSA_generate_key_ex(rsa, RSA_2048_KEY_BITS, bne, nullptr);
                 break;
             case EH_RSA_3072:
-                RSA_generate_key_ex(rsa, RSA_3072_KEY_BITS, bne, nullptr);
+                res = RSA_generate_key_ex(rsa, RSA_3072_KEY_BITS, bne, nullptr);
+                break;
+            default:
+                res = RSA_generate_key_ex(rsa, RSA_3072_KEY_BITS, bne, nullptr);
```
Comment (syan10) [inline core/Enclave/enclave_hsm.cpp:331]: why default to generate 3072?

File: core/Enclave/enclave_hsm.cpp [hunk 11]
```diff
@@ -294,27 +294,59 @@ sgx_status_t enclave_create_rsa_key(uint8_t *cmk_blob,
...
-        PEM_write_bio_RSAPublicKey(public_key, rsa);
+        // storage rsa key with pem format
+        public_key = BIO_new(BIO_s_mem());
+        private_key = BIO_new(BIO_s_mem());
+        if (public_key == NULL || private_key == NULL) {
+            break;
+        }
+
+        PEM_write_bio_RSAPrivateKey(private_key, rsa,
+                                    EVP_des_ede3_ofb(),
```
Comment (syan10) [inline core/Enclave/enclave_hsm.cpp:346]: which algorithm used to encrypt the private key?

File: core/Enclave/enclave_hsm.cpp (hunk 2 above)
Comment (syan10) [reply core/Enclave/enclave_hsm.cpp:496]: you need to check the ciphertext_len by the keyspec.

File: core/Enclave/enclave_hsm.cpp (hunk 8 above)
Comment (syan10) [reply core/Enclave/enclave_hsm.cpp:557]: you should not pass the string to enclave. you need to format the parameter into a data buffer.

File: ehsm_kms_service/delete_secret_thread.js [hunk 12]
```diff
@@ -0,0 +1,118 @@
+const {
+    base64_decode
```
Comment (syan10) [inline ehsm_kms_service/delete_secret_thread.js:2]: why you include the previous commit?

File: core/App/ehsm_provider.cpp (hunk 9 above)
Comment (whh39) [reply core/App/ehsm_provider.cpp:595]: only on line 564, I plan to merge the code for 3072 and 2048

File: core/Enclave/enclave_hsm.cpp (hunk 1 above)
Comment (whh39) [reply core/Enclave/enclave_hsm.cpp:542]: check padding mode and keyspec on line 545

File: core/Enclave/enclave_hsm.cpp (hunk 2 above)
Comment (whh39) [reply core/Enclave/enclave_hsm.cpp:540]: ciphertext_len, keyspec and  cmk_blob_size have been used and ciphertext_len is going to be checked by keyspec

**Pull request closed without merging**